gitlab/resource_gitlab_project_access_token.go

```diff
@@ -0,0 +1,180 @@
+		Read:   resourceGitlabProjectAccessTokenRead,
+		Delete: resourceGitlabProjectAccessTokenDelete,
+
+		Schema: map[string]*schema.Schema{
+			"project": {
+				Type:     schema.TypeInt,
+				Required: true,
+				ForceNew: true,
+			},
+			"name": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"scopes": {
+				Type:     schema.TypeSet,
+				Required: true,
+				ForceNew: true,
+				Elem: &schema.Schema{
+					Type:         schema.TypeString,
+					ValidateFunc: validation.StringInSlice([]string{"api", "read_api", "read_repository", "write_repository"}, false),
+				},
+			},
+			"expires_at": {
+				Type:     schema.TypeString,
+				Optional: true,
+				ValidateFunc: func(i interface{}, k string) (warnings []string, errors []error) {
+					v, ok := i.(string)
+					if !ok {
+						errors = append(errors, fmt.Errorf("expected type of %q to be string", k))
+						return warnings, errors
+					}
+
+					if _, err := time.Parse("2006-01-02", v); err != nil {
+						errors = append(errors, fmt.Errorf("expected %q to be a valid YYYY-MM-DD date, got %q: %+v", k, i, err))
+					}
+
+					return warnings, errors
+				},
+				ForceNew: true,
+			},
+			"token": {
+				Type:      schema.TypeString,
+				Computed:  true,
+				Sensitive: true,
+			},
+			"active": {
+				Type:     schema.TypeBool,
+				Computed: true,
+			},
+			"created_at": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"revoked": {
+				Type:     schema.TypeBool,
+				Computed: true,
+			},
+			"user_id": {
+				Type:     schema.TypeInt,
+				Computed: true,
+			},
+		},
+	}
+}
+
+func resourceGitlabProjectAccessTokenCreate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*gitlab.Client)
+	project := d.Get("project").(int)
+	options := &gitlab.CreateProjectAccessTokenOptions{
+		Name:   gitlab.String(d.Get("name").(string)),
+		Scopes: *stringSetToStringSlice(d.Get("scopes").(*schema.Set)),
+	}
+
+	log.Printf("[DEBUG] create gitlab ProjectAccessToken %s %s", *options.Name, options.Scopes)
+
+	if v, ok := d.GetOk("expires_at"); ok {
+		parsedExpiresAt, err := time.Parse("2006-01-02", v.(string))
+		if err != nil {
+			return fmt.Errorf("Invalid expires_at date: %v", err)
+		}
+		parsedExpiresAtISOTime := gitlab.ISOTime(parsedExpiresAt)
+		options.ExpiresAt = &parsedExpiresAtISOTime
+		log.Printf("[DEBUG] create gitlab ProjectAccessToken with expires_at %s", *options.ExpiresAt)
+	}
+
+	projectAccessToken, _, err := client.ProjectAccessTokens.CreateProjectAccessToken(project, options)
+	if err != nil {
+		return err
+	}
+
+	d.SetId(strconv.Itoa(projectAccessToken.ID))
+	d.Set("token", projectAccessToken.Token)
+
+	return resourceGitlabProjectAccessTokenRead(d, meta)
+}
+
+func resourceGitlabProjectAccessTokenRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*gitlab.Client)
+	project := d.Get("project").(int)
+	projectAccessTokenID, err := strconv.Atoi(d.Id())
```

Comment:
The resource ID format should be `<project-id>:<token-id>`. A good rule of thumb is to include all IDs that you need to pass to the go-gitlab client to GET the object. You should also parse the project ID from the resource ID instead of reading the attribute. This will help you avoid needing a custom import function since you can then use `ImportStatePassthrough`. See `gitlab_project_mirror` for reference:

https://github.com/gitlabhq/terraform-provider-gitlab/blob/8e2faa78634fe687a5d491e3370ca9fd206f2476/gitlab/resource_gitlab_project_mirror.go#L18-L20

https://github.com/gitlabhq/terraform-provider-gitlab/blob/8e2faa78634fe687a5d491e3370ca9fd206f2476/gitlab/resource_gitlab_project_mirror.go#L132-L137

Please remember to add a `## Import` section to the docs as well. 😉

Reply:
No import added as described in the comment above